includes/core/server/class-request.php

```diff
@@ -0,0 +1,523 @@
+		$this->api_client     = $api_client;
+		$this->http_interface = $http_interface;
+	}
+
+	/**
+	 * Returns the needed API.
+	 *
+	 * @return string Check WCPay\Core\Server\APIs.
+	 */
+	abstract public function get_api(): string;
+
+	/**
+	 * Returns the method of the request.
+	 *
+	 * @return string See the constants in WordPress's `Requests` class.
+	 */
+	abstract public function get_method(): string;
+
+	/**
+	 * This is a legacy method, and is the same throughout the codebase.
+	 * Might be worth removing while refactoring to use the Core\Server API.
+	 *
+	 * @return bool
+	 */
+	public function is_site_specific(): bool {
+		return true;
+	}
+
+	/**
+	 * If true, the request will be signed with the user token rather than blog token. Defaults to false.
+	 *
+	 * @return bool
+	 */
+	public function should_use_user_token(): bool {
+		return false;
+	}
+
+	/**
+	 * Indicates if the raw response should be returned.
+	 *
+	 * @return bool
+	 */
+	public function should_return_raw_response(): bool {
+		return false;
+	}
+
+	/**
+	 * Returns all of the parameters for the request.
+	 *
+	 * @return array
+	 * @throws Invalid_Request_Parameter_Exception If the request has not been initialized yet.
+	 */
+	final public function get_params() {
+		$defaults = $this->get_default_params();
+		$params   = array_merge( $defaults, $this->params );
+
+		$missing_params = [];
+		foreach ( $this->get_required_params() as $name ) {
+			if ( ! isset( $params[ $name ] ) ) {
+				$missing_params[] = $name;
+			}
+		}
+
+		if ( ! empty( $missing_params ) ) {
+			throw new Invalid_Request_Parameter_Exception(
+				sprintf(
+					'Trying to access the parameters of a request which is not (fully) initialized yet. Missing parameter(s) for %s: %s',
+					get_class( $this ),
+					implode( ', ', $missing_params )
+				),
+				'wcpay_core_invalid_request_parameter_missing_parameters'
+			);
+		}
+
+		foreach ( $params as $key => $value ) {
+			if ( true === $value ) {
+				// The WCPay server requires the string 'true'.
+				$params[ $key ] = 'true';
+			}
+		}
+
+		return $params;
+	}
+
+	/**
+	 * Get request param by key.
+	 *
+	 * @param string $key Key to get.
+	 *
+	 * @return mixed
+	 * @throws Invalid_Request_Parameter_Exception
+	 */
+	final public function get_param( $key ) {
+		if ( array_key_exists( $key, $this->params ) ) {
+			return $this->params[ $key ];
+		}
+		throw new Invalid_Request_Parameter_Exception( 'The passed key does not exist in Request class', 'wcpay_core_invalid_request_parameter_uninitialized_param' );
+	}
+
+	/**
+	 * Allows the request to be modified, and then sends it.
+	 *
+	 * @param string $hook    The filter to use.
+	 * @param mixed  ...$args Other parameters for the hook.
+	 * @return mixed          Either the response array, or the correct object.
+	 */
+	final public function send( $hook, ...$args ) {
+		$request = $this->apply_filters( $hook, ...$args );
```

Comment:
Do you mean that it requires more context (additional args for the filter)? Otherwise this is a `final` method exactly to prevent child classes from messing with the main request.

Now, the question might be "What if two separate pieces of code try to apply their own `Custom_Request::extend()` method?" We probably need to cover this scenario as well actually, but I think it's doable. @zmaglica could you tinker with this idea?